### PR TITLE
[NN] Refactor the Code Structure of GT

### PR DESCRIPTION
**Pull request by rudongyu**
## Description
1. Refactor the code structure of graph transformer utilities.
2. Fix the attention nan value pollution issue caused by all -inf values in a row of attn_bias (#5476)
    - attn_bias by SpatialEncoder or PathEncoder will be padded with zeros
    - add a note for attn_mask in BiasedMHA to avoid all positions in a row masked
3. Fix scaling factor bug in `BiasedMHA`: change `/ self.scaling` to `* self.scaling`

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by mufeili** [comment]
This PR is quite large. Is it possible to break it into multiple smaller PRs?

**Comment by rudongyu** [comment]
> This PR is quite large. Is it possible to break it into multiple smaller PRs?

Most modifications are just movement of existing utilities for better code structure.
Main changes include:
- rename `LaplacianPosEnc` to `LapPosEncoder`, `BiasedMultiheadAttention` to `BiasedMHA`
- modify initialization of attention bias in `SpatialEncoder` and `PathEncoder` to zero tensors and specify the format of attn_mask in `BiasedMHA` (resolve #5476 )
- documentation fix

File: docs/source/api/python/nn-pytorch.rst [hunk 1]
```diff
@@ -145,3 +145,18 @@ Network Embedding Modules
...
     ~dgl.nn.pytorch.DeepWalk
     ~dgl.nn.pytorch.MetaPath2Vec
+
+Utility Modules for Graph Transformer
+----------------------------------------
+.. autosummary::
+    :toctree: ../../generated/
+    :nosignatures:
+    :template: classtemplate.rst
+
```
Comment (mufeili) [inline docs/source/api/python/nn-pytorch.rst:155]: Is it necessary to have a `gt` namespace? Is this settled from a prior discussion?

File: python/dgl/nn/pytorch/gt/degree_enc.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        ----------
+        g : DGLGraph
+            A DGLGraph to be encoded. If it is a heterogeneous one,
+            it will be transformed into a homogeneous one first.
+
+        Returns
+        -------
+        Tensor
+            Return degree embedding vectors of shape :math:`(N, d)`,
+            where :math:`N` is th number of nodes in the input graph and
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/degree_enc.py:68]:
```suggestion
            where :math:`N` is the number of nodes in the input graph and
```

File: python/dgl/nn/pytorch/gt/degree_enc.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            :math:`d` is :attr:`embedding_dim`.
+        """
+        if len(g.ntypes) > 1 or len(g.etypes) > 1:
+            g = to_homogeneous(g)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/degree_enc.py:72]: You don't necessarily need to address in this PR. I personally prefer performing `to_homogeneous` outside.

File: python/dgl/nn/pytorch/gt/path_enc.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        Returns
+        -------
+        torch.Tensor
+            Return attention bias as path encoding, of shape
+            :math:`(B, N, N, H)`, where :math:`B` is the batch size of
+            the input graph, :math:`N` is the maximum number of nodes, and
+            :math:`H` is :attr:`num_heads`.
+        """
+        device = g.device
+        g_list = unbatch(g)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/path_enc.py:70]: ditto for `unbatch`

File: python/dgl/nn/pytorch/gt/biased_mha.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    Compute attention between nodes with attention bias obtained from graph
+    structures, as introduced in `Do Transformers Really Perform Bad for
+    Graph Representation? <https://arxiv.org/pdf/2106.05234>`__
+
+    .. math::
+
+        \text{Attn}=\text{softmax}(\dfrac{QK^T}{\sqrt{d}} \circ b)
+
+    :math:`Q` and :math:`K` are feature representation of nodes. :math:`d`
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/biased_mha.py:19]:
```suggestion
    :math:`Q` and :math:`K` are feature representations of nodes. :math:`d`
```

File: python/dgl/nn/pytorch/gt/biased_mha.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    is the corresponding :attr:`feat_size`. :math:`b` is attention bias, which
+    can be additive or multiplicative according to the operator :math:`\circ`.
+
+    Parameters
+    ----------
+    feat_size : int
+        Feature size.
+    num_heads : int
+        Number of attention heads, by which attr:`feat_size` is divisible.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/biased_mha.py:28]:
```suggestion
        Number of attention heads, by which :attr:`feat_size` is divisible.
```

File: python/dgl/nn/pytorch/gt/biased_mha.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+            A 3D input tensor. Shape: (batch_size, N, :attr:`feat_size`), where
+            N is the maximum number of nodes.
+        attn_bias : torch.Tensor, optional
+            The attention bias used for attention modification. Shape:
+            (batch_size, N, N, :attr:`num_heads`).
+        attn_mask : torch.Tensor, optional
+            The attention mask used for avoiding computation on invalid
+            positions, where invalid positions are indicated by `True` values.
+            Shape: (batch_size, N, N). Note: For rows corresponding to
+            unexisting nodes, make sure at least one value is set to `False`.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/biased_mha.py:106]:
```suggestion
            unexisting nodes, make sure at least one entry is set to `False` to prevent obtaining NaNs with softmax.
```

File: python/dgl/nn/pytorch/gt/biased_mha.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        y : torch.Tensor
+            The output tensor. Shape: (batch_size, N, :attr:`feat_size`)
+        """
+        q_h = self.q_proj(ndata).transpose(0, 1)
+        k_h = self.k_proj(ndata).transpose(0, 1)
+        v_h = self.v_proj(ndata).transpose(0, 1)
+        bsz, N, _ = ndata.shape
+        q_h = (
+            q_h.reshape(N, bsz * self.num_heads, self.head_dim).transpose(0, 1)
+            * self.scaling
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/biased_mha.py:119]: Why did you change `/ self.scaling` to `* self.scaling`?

File: python/dgl/nn/pytorch/gt/graphormer.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            A 3D input tensor. Shape: (batch_size, N, :attr:`feat_size`), where
+            N is the maximum number of nodes.
+        attn_bias : torch.Tensor, optional
+            The attention bias used for attention modification. Shape:
+            (batch_size, N, N, :attr:`num_heads`).
+        attn_mask : torch.Tensor, optional
+            The attention mask used for avoiding computation on invalid
+            positions, where invalid positions are indicated by `True` values.
+            Shape: (batch_size, N, N). Note: For rows corresponding to
+            unexisting nodes, make sure at least one value is set to `False`.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/graphormer.py:102]: ditto

File: python/dgl/nn/pytorch/gt/spatial_enc.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+from ....batch import unbatch
+from ....transforms import shortest_dist
+
+
+class SpatialEncoder(nn.Module):
+    r"""Spatial Encoder, as introduced in
+    `Do Transformers Really Perform Bad for Graph Representation?
+    <https://proceedings.neurips.cc/paper/2021/file/f1c1592588411002af340cbaedd6fc33-Paper.pdf>`__
+
+    This module is a learnable spatial embedding module which encodes
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/spatial_enc.py:18]:
```suggestion
    This module is a learnable spatial embedding module, which encodes
```

File: python/dgl/nn/pytorch/gt/spatial_enc.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+            of shape :math:`(N, 3)`,
+            where :math:`N`: is the number of nodes in :attr:`g`.
+        node_type : torch.Tensor, optional
+            Node types of :attr:`g`. Default : None.
+
+            * If :attr:`max_node_type` is not 1, :attr:`node_type` needs to
+              be a tensor in shape :math:`(N,)`. The scaling factors of
+              each pair of nodes are determined by their node types.
+            * Otherwise, :attr:`node_type` should be None.
+        Returns
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/spatial_enc.py:179]: I suggest having an empty line above.

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+    n_head : int, optional
+        Number of heads in Transformer Encoder.
+        Default : 1.
+    batch_norm : bool, optional
+        If True, apply batch normalization on raw LaplacianPE.
+        Default : False.
+    num_post_layer : int, optional
+        If num_post_layer > 0, apply an MLP of ``num_post_layer`` layers after
+        pooling.
+        Default : 0.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/lap_enc.py:34]: Merge L33 and 34?

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+    Example
+    -------
+    >>> import dgl
+    >>> from dgl import LaplacianPE
+    >>> from dgl.nn import LapPosEncoder
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/lap_enc.py:40]: Why did you want to break the consistency of using `Laplacian` for `PE` and `PosEncoder`?

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+    >>> transform = LaplacianPE(k=5, feat_name='eigvec', eigval_name='eigval', padding=True)
+    >>> g = dgl.graph(([0,1,2,3,4,2,3,1,4,0], [2,3,1,4,0,0,1,2,3,4]))
+    >>> g = transform(g)
+    >>> eig_vals, eig_vecs = g.ndata['eigval'], g.ndata['eigvec']
+    >>> TransformerLPE = LapPosEncoder(model_type="Transformer", num_layer=3, k=5,
```
Comment (mufeili) [inline python/dgl/nn/pytorch/gt/lap_enc.py:46]: Use lowercase for object names?

**Review by mufeili — COMMENTED** [review]
I've done a pass. For a sanity check, did you try the refactored code on an end-to-end example?

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+            self.post_mlp = nn.Sequential(*layers)
+        else:
+            self.post_mlp = None
+
+    def forward(self, eig_vals, eig_vecs):
+        r"""
+        Parameters
+        ----------
+        eig_vals : Tensor
+            Laplacian Eigenvalues of shape :math:`(N, k)`, k different eigenvalues repeat N times,
```
Comment (frozenbugs) [inline python/dgl/nn/pytorch/gt/lap_enc.py:121]: line too long

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+            can be obtained by using `LaplacianPE`.
+        eig_vecs : Tensor
+            Laplacian Eigenvectors of shape :math:`(N, k)`, can be obtained by using `LaplacianPE`.
```
Comment (frozenbugs) [inline python/dgl/nn/pytorch/gt/lap_enc.py:124]: ditto

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+                for _ in range(num_post_layer - 2):
+                    layers.append(nn.Linear(2 * lpe_dim, 2 * lpe_dim))
+                    layers.append(nn.ReLU())
+                layers.append(nn.Linear(2 * lpe_dim, lpe_dim))
+                layers.append(nn.ReLU())
+            self.post_mlp = nn.Sequential(*layers)
+        else:
+            self.post_mlp = None
+
+    def forward(self, eig_vals, eig_vecs):
```
Comment (frozenbugs) [inline python/dgl/nn/pytorch/gt/lap_enc.py:116]: eigvals, eigvecs

eigenvalue and eigenvector are single word instead 2 words, not underscore required.

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    ):
+        super(LapPosEncoder, self).__init__()
+        self.model_type = model_type
+        self.linear = nn.Linear(2, lpe_dim)
+
+        if self.model_type == "Transformer":
+            encoder_layer = nn.TransformerEncoderLayer(
+                d_model=lpe_dim, nhead=n_head, batch_first=True
+            )
+            self.pe_encoder = nn.TransformerEncoder(
```
Comment (frozenbugs) [inline python/dgl/nn/pytorch/gt/lap_enc.py:72]: what does pe in pe_encoder stand for?

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        pos_enc = self.linear(pos_enc)
+
+        if self.model_type == "Transformer":
+            pos_enc = self.pe_encoder(
+                src=pos_enc, src_key_padding_mask=empty_mask[:, :, 1]
+            )
+        else:
+            pos_enc = self.pe_encoder(pos_enc)
+
+        # Remove masked sequences
```
Comment (frozenbugs) [inline python/dgl/nn/pytorch/gt/lap_enc.py:150]: nit: end with dot.

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+        pos_enc[empty_mask[:, :, 1]] = 0
+
+        # Sum pooling
```
Comment (frozenbugs) [inline python/dgl/nn/pytorch/gt/lap_enc.py:153]: ditto

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+        pos_enc = th.sum(pos_enc, 1, keepdim=False)
+
+        # MLP post pooling
```
Comment (frozenbugs) [inline python/dgl/nn/pytorch/gt/lap_enc.py:156]: ditto.

**Review by frozenbugs — COMMENTED** [review]
Done a pass, there are still 2 files I did not review, please check according to the comment on the existing files.

File: python/dgl/nn/pytorch/gt/degree_enc.py (hunk 3 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/degree_enc.py]: agree.

File: docs/source/api/python/nn-pytorch.rst (hunk 1 above)
Comment (rudongyu) [reply docs/source/api/python/nn-pytorch.rst:155]: We previously decided to move gt-related modules to a separate folder for better management. Besides, in my opinion, maintaining a separate namespace avoids a file too large and gives clear guidance for contributors.

File: python/dgl/nn/pytorch/gt/biased_mha.py (hunk 8 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/biased_mha.py:120]: That's a bug found in our pre-training experiments.

File: python/dgl/nn/pytorch/gt/lap_enc.py (hunk 13 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/lap_enc.py:39]: My consideration here is to keep consistency with `SpatialEncoder`, `DegreeEncoder`, etc. However, `LaplacianPosEncoder` seems too long. Any suggestion?

File: python/dgl/nn/pytorch/gt/biased_mha.py (hunk 8 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/gt/biased_mha.py:120]: perhaps worth documenting it in the PR description in case we need to reference it in the future

File: python/dgl/nn/pytorch/gt/lap_enc.py (hunk 13 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/gt/lap_enc.py:39]: Alternatively, you can rename `LaplacianPE` to `LapPE`. To maintain backward compatibility, consider having an alias for now and adding a deprecation warning.

File: python/dgl/nn/pytorch/gt/lap_enc.py [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+            Laplacian Eigenvectors of shape :math:`(N, k)`, can be obtained by using `LaplacianPE`.
+
+        Returns
+        -------
+        Tensor
+            Return the laplacian positional encodings of shape :math:`(N, d)`,
+            where :math:`N` is the number of nodes in the input graph,
+            :math:`d` is :attr:`lpe_dim`.
+        """
+        pos_enc = th.cat(
```
[the remark this answers is not in the document]
Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/lap_enc.py]: modified to `pos_encoding`

File: python/dgl/nn/pytorch/gt/lap_enc.py (hunk 18 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/lap_enc.py]: pe stands for positional encoding, which is a common abbr. in this context.

File: python/dgl/nn/pytorch/gt/degree_enc.py [hunk 23]
```diff
@@ -0,0 +1,89 @@
+"""Degree Encoder"""
+
+import torch as th
+import torch.nn as nn
+
+from .... import to_homogeneous
```
[the remark this answers is not in the document]
Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/degree_enc.py]: Absolute path is indeed more recommended than relative import. Currently relative import is a convention. Maybe we need further discussion to refactor all the code.

**Review by mufeili — APPROVED** [review]
I'm good.

**Review by frozenbugs — APPROVED** [review]
Please apply the lint suggestion.

**Comment by rudongyu** [comment]
@dgl-bot

**Comment by rudongyu** [comment]
@dgl-bot
